jwst/saturation/saturation.py

```diff
@@ -160,6 +165,31 @@ def irs2_flag_saturation(input_model, ref_model, n_pix_grow_sat):
                                         irs2_mask, detector)
             # check for saturation
             flag_temp = np.where(sci_temp >= sat_thresh, SATURATED, 0)
+            if group == 2:
+                # Identify groups which we wouldn't expect to saturate by the third group,
+                # on the basis of the first group
+                scigp1 = x_irs2.from_irs2(data[ints, 0, :, :], irs2_mask, detector)
+                mask = scigp1 / np.mean(read_pattern[0]) * read_pattern[2][-1] < sat_thresh
+
+                # Identify groups with suspiciously large values in the second group
+                scigp2 = x_irs2.from_irs2(data[ints, 1, :, :], irs2_mask, detector)
+                mask &= scigp2 > sat_thresh / len(read_pattern[1])
+
+                # Identify groups that are saturated in the third group
+                gp3mask = np.where(flag_temp & SATURATED, True, False)
+                mask &= gp3mask
+
+                # Flag the 2nd group for the pixels passing that gauntlet
+                dq_temp = x_irs2.from_irs2(groupdq[ints, 1, :, :], irs2_mask, detector)
```

Comment:
Nearly there, but I think this line is problematic.  Since it's using the existing groupdq array, it's going to be growing ALL pixels flagged as SATURATED, not just those that passed the gauntlet above, so things in group1 will get grown twice.

I think this should be possible to deal with by replacing 
dq_temp = x_irs2.from_irs2(groupdq[ints, 1, :, :], irs2_mask, detector)
with
dq_temp = np.zeros_like(mask).astype(int)

Reply:
I suppose it would probably be better form though to explicitly cast the dq_temp to be the same dtype as groupdq.